File: app/src/main/java/wangdaye/com/geometricweather/basic/model/weather/Precipitation.java
package wangdaye.com.geometricweather.basic.model.weather;

import android.content.Context;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import wangdaye.com.geometricweather.R;

/**
 * Precipitation.
 *
 * default unit : {@link wangdaye.com.geometricweather.basic.model.option.unit.PrecipitationUnit#MM}
 * */
public class Precipitation {

    @Nullable private Float total;
    @Nullable private Float thunderstorm;
    @Nullable private Float rain;
    @Nullable private Float snow;
    @Nullable private Float ice;

    public static final float PRECIPITATION_LIGHT = 10;
    public static final float PRECIPITATION_MIDDLE = 25;
    public static final float PRECIPITATION_HEAVY = 50;
    public static final float PRECIPITATION_RAINSTORM = 100;

    public Precipitation(@Nullable Float total,
                         @Nullable Float thunderstorm,
                         @Nullable Float rain,
                         @Nullable Float snow,
                         @Nullable Float ice) {
        this.total = total;
        this.thunderstorm = thunderstorm;
        this.rain = rain;
        this.snow = snow;
        this.ice = ice;
    }

    @Nullable
    public Float getTotal() {
        return total;
    }

    @Nullable
    public Float getThunderstorm() {
        return thunderstorm;
    }

    @Nullable
    public Float getRain() {
        return rain;
    }

    @Nullable
    public Float getSnow() {
        return snow;
    }

    @Nullable
    public Float getIce() {
        return ice;
    }

    public boolean isValid() {
        return total != null && total > 0;
    }

    @ColorInt
    public int getPrecipitationColor(Context context) {
        if (total == null) {
            return ContextCompat.getColor(context, R.color.colorLevel_1);
        } else if (total <= PRECIPITATION_LIGHT) {
            return ContextCompat.getColor(context, R.color.colorLevel_1);
        } else if (total <= PRECIPITATION_MIDDLE) {
            return ContextCompat.getColor(context, R.color.colorLevel_2);
        } else if (total <= PRECIPITATION_HEAVY) {
            return ContextCompat.getColor(context, R.color.colorLevel_3);
        } else if (total <= PRECIPITATION_RAINSTORM) {
            return ContextCompat.getColor(context, R.color.colorLevel_4);
        } else {
            return ContextCompat.getColor(context, R.color.colorLevel_5);
        }
    }
}

File: app/src/main/java/org/breezyweather/common/basic/models/weather/Precipitation.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.basic.models.weather

import android.content.Context
import android.graphics.Color
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import org.breezyweather.R
import java.io.Serializable

/**
 * Precipitation.
 *
 * default unit : [PrecipitationUnit.MM]
 */
class Precipitation(
    val total: Float? = null,
    val thunderstorm: Float? = null,
    val rain: Float? = null,
    val snow: Float? = null,
    val ice: Float? = null
) : Serializable {

    companion object {
        // TODO: Seems a bit high, should probably lower that
        const val PRECIPITATION_LIGHT = 10f
        const val PRECIPITATION_MIDDLE = 25f
        const val PRECIPITATION_HEAVY = 50f
        const val PRECIPITATION_RAINSTORM = 100f
    }

    @ColorInt
    fun getPrecipitationColor(context: Context): Int {
        return if (total == null) {
            Color.TRANSPARENT
        } else when (total) {
            in 0f.. PRECIPITATION_LIGHT -> ContextCompat.getColor(context, R.color.colorLevel_1)
            in PRECIPITATION_LIGHT.. PRECIPITATION_MIDDLE -> ContextCompat.getColor(context, R.color.colorLevel_2)
            in PRECIPITATION_MIDDLE.. PRECIPITATION_HEAVY -> ContextCompat.getColor(context, R.color.colorLevel_3)
            in PRECIPITATION_HEAVY.. PRECIPITATION_RAINSTORM -> ContextCompat.getColor(context, R.color.colorLevel_4)
            in PRECIPITATION_RAINSTORM.. Float.MAX_VALUE -> ContextCompat.getColor(context, R.color.colorLevel_5)
            else -> Color.TRANSPARENT
        }
    }

    val isValid: Boolean
        get() = total != null && total > 0
}
